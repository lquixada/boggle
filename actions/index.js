export const addAttempt = (word, scored) => ({
  type: 'ADD_ATTEMPT',
  word,
  scored
});

export const incrementCounter = () => ({
  type: 'INCREMENT_COUNTER'
});

export const startGame = () => ({
  type: 'START_GAME'
});

export const stopGame = () => ({
  type: 'STOP_GAME'
});
