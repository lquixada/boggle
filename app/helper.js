/*
 * Helper
 */

define(['underscore'], function (_) {
  'use strict';
  
  _.mixin({
    render: function (elementId, context) {
      var source   = $(elementId+'-template').html();
      var compiled = _.template(source);
      var html = compiled(context);

      $(elementId).html(html);
    }
  });
});
