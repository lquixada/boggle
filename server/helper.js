import webpack from 'webpack';
import config from '../webpack.dev.config';

export const compiler = webpack(config);
