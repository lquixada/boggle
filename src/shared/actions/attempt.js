import {isOnScoreList, isOnBoard, isOnDictionary} from '../helpers'

export const addAttempt = (word, scored) => ({
  type: 'ADD_ATTEMPT',
  payload: {
    word,
    scored
  }
})

export const addCheckedAttempt = (word) => async (dispatch, getState) => {
  let onDictionary = false
  const state = getState()
  word = word.toUpperCase()

  const [onBoard, onScoreList] = await Promise.all([
    isOnBoard(state.matrix, word),
    isOnScoreList(state.attempts, word)
  ])

  if (onBoard && !onScoreList) {
    onDictionary = await isOnDictionary(word)
  }

  return dispatch(addAttempt(word, onDictionary))
}
