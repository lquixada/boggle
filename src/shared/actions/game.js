export const startGame = () => ({
  type: 'START_GAME'
});

export const stopGame = () => ({
  type: 'STOP_GAME'
});
