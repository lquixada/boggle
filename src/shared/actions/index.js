/* All actions must be FSA-compliant (https://github.com/acdlite/flux-standard-action). */
export * from './attempt';
export * from './game';
