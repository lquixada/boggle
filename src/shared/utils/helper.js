import {fromJS} from 'immutable';
import fetch from 'cross-fetch';

import config from '../../shared/config';
import BoardChecker from '../utils/board-checker';

export const isOnScoreList = (attempts, word) => {
  const found = attempts.find((attempt) => attempt.get('word') === word);

  return Boolean(found);
};

export const isOnBoard = (matrix, word) => {
  const board = new BoardChecker(matrix);
  return board.hasWord(word);
};

export const isOnDictionary = (word) => {
  const url = `${config.url.wiktionary}&titles=${word.toLowerCase()}`;

  return fetch(url)
    .then((response) => response.json())
    // The "-1" property means "word not found"
    .then((data) => !data.query.pages[-1]);
};

// Abstraction to handle pre-composed state received from server
// (ie, leave top level keys untouched)
export const immutifyState = (obj) => {
  const objMut = {};

  fromJS(obj).forEach((v, k) => {
    objMut[k] = v;
  });

  return objMut;
};
