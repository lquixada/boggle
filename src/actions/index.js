import { isOnScoreList, isOnBoard, isOnDictionary } from '../utils/helper';

export const addAttempt = (word, scored) => ({
  type: 'ADD_ATTEMPT',
  word,
  scored
});

export const startGame = () => ({
  type: 'START_GAME'
});

export const stopGame = () => ({
  type: 'STOP_GAME'
});

export const addCheckedAttempt = (word) => {
  word = word.toUpperCase();

  return (dispatch, getState) => {
    const state = getState();

    return Promise.all([
      !isOnScoreList(state.attempts, word),
      isOnBoard(state.matrix, word),
      isOnDictionary(word)
    ]).then((values) => {
      const result = values.every(v => v);
      return dispatch(addAttempt(word, result));
    });
  };
};
