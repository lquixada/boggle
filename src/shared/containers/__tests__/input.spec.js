import React from 'react';
import {bindActionCreators} from 'redux';
import {mountConnected} from '../../../__tests__/helper';
import configureStore from '../../store';
import * as actionCreators from '../../actions';
import InputContainer from '../input';

describe('<InputContainer />', () => {
  let actions;
  let component;
  let store;

  beforeEach(() => {
    store = configureStore();
    component = mountConnected(<InputContainer />, store);
    actions = bindActionCreators(actionCreators, store.dispatch);
  });

  it('is disabled by default', () => {
    expect(component.find('input').prop('disabled')).toBe(true);
  });

  it('is enabled when the game has started', () => {
    actions.startGame();
    expect(component.update().find('input').prop('disabled')).toBe(false);
  });

  it('is disabled when the game has stopped', () => {
    actions.stopGame();
    expect(component.update().find('input').prop('disabled')).toBe(true);
  });
});
