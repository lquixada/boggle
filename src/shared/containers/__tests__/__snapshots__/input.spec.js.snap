// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputContainer /> is disabled by default 1`] = `
<InputContainer
  started={false}
>
  <Input
    onChange={[Function]}
    onEnter={[Function]}
    onSubmit={[Function]}
    started={false}
    value=""
  >
    <styled.div>
      <StyledComponent
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-hMqMXs",
              "isStatic": true,
              "lastClassName": "bcPEYZ",
              "rules": Array [
                "
  ",
                "
  flex: 1 1 auto;
  -ms-flex: 1 1 auto;
",
                "
  margin-left: 2rem;

  @media (max-width: 540px) {
    ",
                "
  flex: 1 1 100%;
  -ms-flex: 1 1 100%;
",
                "
    ",
                "
  order: 2;
  -ms-flex-order: 2;
",
                "
    margin: 1rem 0 0;
  }
",
              ],
            },
            "displayName": "styled.div",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-hMqMXs",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
      >
        <div
          className="sc-hMqMXs bcPEYZ"
        >
          <styled.input
            disabled={true}
            onChange={[Function]}
            onKeyUp={[Function]}
            placeholder="Press start to begin..."
            type="text"
            value=""
          >
            <StyledComponent
              disabled={true}
              forwardedComponent={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-kEYyzF",
                    "isStatic": true,
                    "lastClassName": "dcKvZB",
                    "rules": Array [
                      "
  ",
                      "
  background-color: #404040;
  border: .1rem solid;
  border-radius: .5rem;
  border-color: #303030 #505050 #505050 #303030;
",
                      "
  width: 100%;
  padding: 1rem;
  color: #fff;
  font-size: 2rem;
  border-radius: .4rem;

  &:disabled {
    opacity: .5;
  }
",
                    ],
                  },
                  "displayName": "styled.input",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-kEYyzF",
                  "target": "input",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                }
              }
              forwardedRef={
                Object {
                  "current": <input
                    class="sc-kEYyzF dcKvZB"
                    disabled=""
                    placeholder="Press start to begin..."
                    type="text"
                    value=""
                  />,
                }
              }
              onChange={[Function]}
              onKeyUp={[Function]}
              placeholder="Press start to begin..."
              type="text"
              value=""
            >
              <input
                className="sc-kEYyzF dcKvZB"
                disabled={true}
                onChange={[Function]}
                onKeyUp={[Function]}
                placeholder="Press start to begin..."
                type="text"
                value=""
              />
            </StyledComponent>
          </styled.input>
        </div>
      </StyledComponent>
    </styled.div>
  </Input>
</InputContainer>
`;

exports[`<InputContainer /> is disabled when the game has stopped 1`] = `
<InputContainer
  started={false}
>
  <Input
    onChange={[Function]}
    onEnter={[Function]}
    onSubmit={[Function]}
    started={false}
    value=""
  >
    <styled.div>
      <StyledComponent
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-hMqMXs",
              "isStatic": true,
              "lastClassName": "bcPEYZ",
              "rules": Array [
                "
  ",
                "
  flex: 1 1 auto;
  -ms-flex: 1 1 auto;
",
                "
  margin-left: 2rem;

  @media (max-width: 540px) {
    ",
                "
  flex: 1 1 100%;
  -ms-flex: 1 1 100%;
",
                "
    ",
                "
  order: 2;
  -ms-flex-order: 2;
",
                "
    margin: 1rem 0 0;
  }
",
              ],
            },
            "displayName": "styled.div",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-hMqMXs",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
      >
        <div
          className="sc-hMqMXs bcPEYZ"
        >
          <styled.input
            disabled={true}
            onChange={[Function]}
            onKeyUp={[Function]}
            placeholder="Press start to begin..."
            type="text"
            value=""
          >
            <StyledComponent
              disabled={true}
              forwardedComponent={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-kEYyzF",
                    "isStatic": true,
                    "lastClassName": "dcKvZB",
                    "rules": Array [
                      "
  ",
                      "
  background-color: #404040;
  border: .1rem solid;
  border-radius: .5rem;
  border-color: #303030 #505050 #505050 #303030;
",
                      "
  width: 100%;
  padding: 1rem;
  color: #fff;
  font-size: 2rem;
  border-radius: .4rem;

  &:disabled {
    opacity: .5;
  }
",
                    ],
                  },
                  "displayName": "styled.input",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-kEYyzF",
                  "target": "input",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                }
              }
              forwardedRef={
                Object {
                  "current": <input
                    class="sc-kEYyzF dcKvZB"
                    disabled=""
                    placeholder="Press start to begin..."
                    type="text"
                    value=""
                  />,
                }
              }
              onChange={[Function]}
              onKeyUp={[Function]}
              placeholder="Press start to begin..."
              type="text"
              value=""
            >
              <input
                className="sc-kEYyzF dcKvZB"
                disabled={true}
                onChange={[Function]}
                onKeyUp={[Function]}
                placeholder="Press start to begin..."
                type="text"
                value=""
              />
            </StyledComponent>
          </styled.input>
        </div>
      </StyledComponent>
    </styled.div>
  </Input>
</InputContainer>
`;

exports[`<InputContainer /> is enabled when the game has started 1`] = `
<InputContainer
  started={true}
>
  <Input
    onChange={[Function]}
    onEnter={[Function]}
    onSubmit={[Function]}
    started={true}
    value=""
  >
    <styled.div>
      <StyledComponent
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-hMqMXs",
              "isStatic": true,
              "lastClassName": "bcPEYZ",
              "rules": Array [
                "
  ",
                "
  flex: 1 1 auto;
  -ms-flex: 1 1 auto;
",
                "
  margin-left: 2rem;

  @media (max-width: 540px) {
    ",
                "
  flex: 1 1 100%;
  -ms-flex: 1 1 100%;
",
                "
    ",
                "
  order: 2;
  -ms-flex-order: 2;
",
                "
    margin: 1rem 0 0;
  }
",
              ],
            },
            "displayName": "styled.div",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-hMqMXs",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
      >
        <div
          className="sc-hMqMXs bcPEYZ"
        >
          <styled.input
            disabled={false}
            onChange={[Function]}
            onKeyUp={[Function]}
            placeholder="Type the word and hit Enter!"
            type="text"
            value=""
          >
            <StyledComponent
              disabled={false}
              forwardedComponent={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-kEYyzF",
                    "isStatic": true,
                    "lastClassName": "dcKvZB",
                    "rules": Array [
                      "
  ",
                      "
  background-color: #404040;
  border: .1rem solid;
  border-radius: .5rem;
  border-color: #303030 #505050 #505050 #303030;
",
                      "
  width: 100%;
  padding: 1rem;
  color: #fff;
  font-size: 2rem;
  border-radius: .4rem;

  &:disabled {
    opacity: .5;
  }
",
                    ],
                  },
                  "displayName": "styled.input",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-kEYyzF",
                  "target": "input",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                }
              }
              forwardedRef={
                Object {
                  "current": <input
                    class="sc-kEYyzF dcKvZB"
                    placeholder="Type the word and hit Enter!"
                    type="text"
                    value=""
                  />,
                }
              }
              onChange={[Function]}
              onKeyUp={[Function]}
              placeholder="Type the word and hit Enter!"
              type="text"
              value=""
            >
              <input
                className="sc-kEYyzF dcKvZB"
                disabled={false}
                onChange={[Function]}
                onKeyUp={[Function]}
                placeholder="Type the word and hit Enter!"
                type="text"
                value=""
              />
            </StyledComponent>
          </styled.input>
        </div>
      </StyledComponent>
    </styled.div>
  </Input>
</InputContainer>
`;
